### ISSUE TITLE
X.tryParse inconsitencies

### ISSUE THREAD

**Issue by osundblad** [google/guava#2238]
The `tryParse` methods should behave similarly: 
```
Doubles.tryParse(null)  -> null
Floats.tryParse(null)   -> null
Longs.tryParse(null)    -> NPE
Ints.tryParse(null)     -> NPE
```
For a method like this I would prefer the null result since otherwise I would use the Java parseInt. But if they all should throw NPE I think that should be clearer in the JavaDoc (even if you normally do not document @NotNull parameters).

**Comment by kluever** [comment]
These should all throw a NPE...not sure why that's not the case. It seems like `Doubles.tryParse()` should throw:
```
  public static Double tryParse(String string) {
    if (FLOATING_POINT_PATTERN.matcher(string).matches()) {
```

But maybe that returns false?

**Comment by osundblad** [comment]
Just an idea: maybe they should return an optional and never throw? But then that would really break the api.

**Comment by kluever** [comment]
I just added the following tests, and these tests all pass (i.e., they all throw `NullPointerException`):

```
  public void testTryParse_withNullGwt() {
    try {
      long unused = Longs.tryParse(null);
      fail("Expected NPE");
    } catch (NullPointerException expected) {
    }
    try {
      int unused = Ints.tryParse(null);
      fail("Expected NPE");
    } catch (NullPointerException expected) {
    }
  }

  @GwtIncompatible
  public void testTryParse_withNullNoGwt() {
    try {
      double unused = Doubles.tryParse(null);
      fail("Expected NPE");
    } catch (NullPointerException expected) {
    }
    try {
      float unused = Floats.tryParse(null);
      fail("Expected NPE");
    } catch (NullPointerException expected) {
    }
    try {
      long unused = Longs.tryParse(null);
      fail("Expected NPE");
    } catch (NullPointerException expected) {
    }
    try {
      int unused = Ints.tryParse(null);
      fail("Expected NPE");
    } catch (NullPointerException expected) {
    }
  }
```

@osundblad Where are you seeing the null -> null behavior?

**Comment by Quurks** [comment]
The String "null" leads to null - but for all tryParse variations (and its expected).

**Comment by kluever** [comment]
@Quurks I don't think this discussion is about the string `"null"`, but rather the value `null`.

**Comment by Quurks** [comment]
@kluever Yes, but the String "null" could be (part of) the reason @osundblad is experiencing his results, while your tests pass, "null" and null look similar in a logfile.

**Comment by kluever** [comment]
To be clear: `tryParse("null")` returns `null` for all of the methods in question.

**Comment by osundblad** [comment]
No it is not the String null it is exacyly as I wrote it at the top:

    Doubles.tryParse(null)  -> null
    Floats.tryParse(null)   -> null
    Longs.tryParse(null)    -> NPE
    Ints.tryParse(null)     -> NPE

And since you usually use methods like this in I/O I think they should be very forgiving, ie. accept null and not throw NPE on it. 

I found it while trying to convert json to Java Objects where the JavaScript framework inserted null values instead of empty string (the json included both Integers and Doubles).

**Comment by AlihassanProgrammer** [comment]
just simply null the string

**Comment by kluever** [comment]
@osundblad I still don't understand how you're getting `null` back from `Doubles.tryParse(null)`. What version of Guava are you using?

Please see my test cases above. The methods are behaving exactly as they should (they throw a `NullPointerException` when passed `null`).

**Comment by Quurks** [comment]
The testcase is using `long unused = Longs.tryParse(null);` which is problematic: 
`Longs.tryParse` returns a boxed `Long`, which is unboxed to the primitive `long`.
Unboxing a null value yields a NPE, so even something returning `null`, for instance `Longs.tryParse("a")` passes the test.
Solution: use the boxed type `Long` for the `unused` var

**Comment by lowasser** [comment]
@Quurks we've fixed the test, and we're still correctly getting an NPE for each of those cases.  (We haven't mirrored it out yet, though.)

**Comment by lowasser** [comment]
Based on our tests resulting in correct, consistent behavior, marking this obsolete.

**Comment by osundblad** [comment]
Sorry for the late reply. I tried to figure out how I got the null result from Doubles but could not recreate it in the hour I spent. Kind of annoyed since I and a co-worker were pair-programming and wrote JUnit test tests to prove the behavior (the I actually copy/pasted the results from the tests) at the time but since we wanted null not NPE we switched from guava to apache commons so we never checked in the tests. 

The only things I can think of is that it was an:
1 We used an old version of guava (tried a few without being able to recreate the weird behavior)
2 We somehow got another library without noticing (the dependency list was quit long)

But it is all the same since the behavior is consistent (now). (But sadly not what we needed so were using apache commons)

Sorry for the miss report
